lib/ecto/query.ex

```diff
@@ -2262,6 +2262,49 @@ defmodule Ecto.Query do
     has_named_binding?(Ecto.Queryable.to_query(queryable), key)
   end
 
+  @doc """
+  Applies the callback function to the query if it does not have a named binding,
+  otherwise returns the query without changes.
+
+  The function accepts an `Ecto.Query` struct and must return the query with an added 
+  named binding.
+
+  ## Examples
+
+      with_named_binding(query, :comments, fn query ->
+        join(query, :left, c in assoc(p, :comments), as: :comments)
+      end)
+
+  For more information on named bindings see "Named bindings" in this module doc.
+  """
+  def with_named_binding(%Ecto.Query{} = query, key, fun) when is_function(fun, 1) do
+    if has_named_binding?(query, key) do
+      query
+    else
+      query
+      |> fun.()
```

Comment:
```suggestion
  def with_named_binding(%Ecto.Query{} = query, key, fun) when is_function(fun, 2) do
    if has_named_binding?(query, key) do
      query
    else
      fun.(query, key)
```

This was one of @v0idpwn's suggestion in the proposal discussion. It's a really good idea because then you can create reusable callbacks.

Reply:
We can also check the arity and support both.

Reply:
I didn't comment in the issue but I got the inspiration from `validate_change`, whose typespec is:
```elixir
@spec validate_change(
  t(), 
  atom(), 
  (atom(), term() -> [{atom(), String.t()} | {atom(), {String.t(), Keyword.t()}}])
) :: t()
```